internal/kubex/config.go

```diff
@@ -35,53 +37,67 @@ type ConfigWithMeta struct {
-	kubeconfigPath := os.Getenv(clientcmd.RecommendedConfigPathEnvVar)
-	if len(kubeconfigPath) == 0 {
-		if c, err := rest.InClusterConfig(); err == nil {
-			return &ConfigWithMeta{
-				K8s:            c,
-				CurrentContext: "In cluster",
-			}, nil
+		loadingRules.ExplicitPath = kubeconfig
+	} else {
+		// 2. KUBECONFIG environment variable pointing at a file
+		kubeconfigPath := os.Getenv(clientcmd.RecommendedConfigPathEnvVar)
+		if len(kubeconfigPath) == 0 {
+			// 3. In-cluster config if running in cluster
+			if c, err := rest.InClusterConfig(); err == nil {
+				return &ConfigWithMeta{
+					K8s:            c,
+					CurrentContext: "In cluster",
+				}, nil
+			}
+		} else {
+			loadingRules.ExplicitPath = kubeconfigPath
+			// 4. $HOME/.kube/config if exists
+			// 5. user.HomeDir/.kube/config if exists
+			//
+			// NOTE: For default config file locations, upstream only checks
+			// $HOME for the user's home directory, but we can also try
+			// os/user.HomeDir when $HOME is unset.
+			if _, ok := os.LookupEnv("HOME"); !ok {
+				u, err := user.Current()
+				if err != nil {
+					return nil, fmt.Errorf("could not get current user: %w", err)
+				}
+				loadingRules.Precedence = append(loadingRules.Precedence, filepath.Join(u.HomeDir, clientcmd.RecommendedHomeDir, clientcmd.RecommendedFileName))
+			}
 		}
 	}
 
-	// 3. In-cluster config if running in cluster
-	// 4. $HOME/.kube/config if exists
-	// 5. user.HomeDir/.kube/config if exists
-	//
-	// NOTE: For default config file locations, upstream only checks
-	// $HOME for the user's home directory, but we can also try
-	// os/user.HomeDir when $HOME is unset.
-	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
-	if _, ok := os.LookupEnv("HOME"); !ok {
-		u, err := user.Current()
-		if err != nil {
-			return nil, fmt.Errorf("could not get current user: %w", err)
-		}
-		loadingRules.Precedence = append(loadingRules.Precedence, filepath.Join(u.HomeDir, clientcmd.RecommendedHomeDir, clientcmd.RecommendedFileName))
+	//  1. --kubecontext flag
+	if kubecontext == "" {
+		// 2. KUBECONTEXT env
+		kubecontext = os.Getenv("KUBECONTEXT")
 	}
 
-	return transform(clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, nil))
+	configOverrides := &clientcmd.ConfigOverrides{CurrentContext: kubecontext}
+	return transform(clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides))
 }
 
 func transform(c clientcmd.ClientConfig) (*ConfigWithMeta, error) {
 	rawConfig, err := c.RawConfig()
 	if err != nil {
 		return nil, fmt.Errorf("while getting raw config: %v", err)
 	}
+
 	clientConfig, err := c.ClientConfig()
 	if err != nil {
 		return nil, fmt.Errorf("while getting client config: %v", err)
 	}
+
+	// 3. load from rawConfig
+	if len(kubecontext) == 0 {
+		kubecontext = rawConfig.CurrentContext
+	}
+
 	return &ConfigWithMeta{
 		K8s:            clientConfig,
-		CurrentContext: rawConfig.CurrentContext,
+		CurrentContext: kubecontext,
```

Comment:
This change is not needed, right? As you pass the `&clientcmd.ConfigOverrides{CurrentContext: kubecontext}`, it is already properly set? So that means we can revert this to:

```go
func transform(c clientcmd.ClientConfig) (*ConfigWithMeta, error) {
	rawConfig, err := c.RawConfig()
	if err != nil {
		return nil, fmt.Errorf("while getting raw config: %v", err)
	}

	clientConfig, err := c.ClientConfig()
	if err != nil {
		return nil, fmt.Errorf("while getting client config: %v", err)
	}

	return &ConfigWithMeta{
		K8s:            clientConfig,
		CurrentContext: rawConfig.CurrentContext,
	}, nil
}
```

Reply:
Yes, I also feel like it was redundant.

I was expecting the same but I don't know why the value of `c.RawConfig().CurrentContext` is not equal to the value set by `configOverrides := &clientcmd.ConfigOverrides{CurrentContext: kubecontext}` by the caller of `transform`.

The value of `c.RawConfig().CurrentContext` remains the current context in my `~/.kube/config`, it is not override yet?! 

---
I also tried to configure only `ConfigWithMeta.CurrentContext` in the `transform` function and gave up the overrides in `LoadRestConfigWithMetaInformation`. The `ConfigWithMeta.CurrentContext` was set correctly but Helm was trying to connect to the cluster which was corresponding to the previously context value.

```
	return transform(clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, nil))
}

func transform(c clientcmd.ClientConfig) (*ConfigWithMeta, error) {
	rawConfig, err := c.RawConfig()
	if err != nil {
		return nil, fmt.Errorf("while getting raw config: %v", err)
	}

	clientConfig, err := c.ClientConfig()
	if err != nil {
		return nil, fmt.Errorf("while getting client config: %v", err)
	}

	//  1. --kubecontext flag
	if kubecontext == "" {
		// 2. KUBECONTEXT env
		kubecontext = os.Getenv("KUBECONTEXT")
		// 3. load from rawConfig
		if len(kubecontext) == 0 {
			kubecontext = rawConfig.CurrentContext
		}
	}

	return &ConfigWithMeta{
		K8s:            clientConfig,
		CurrentContext: kubecontext,
	}, nil
}
```

---

I cannot get it work otherwise. I wish to understand but I may not have enough XP with Go. It just let me fell like a pointer issue or I miss understand how overrides work. I'm baffled!

Reply:
Ok, I understand now 👍 Fair enough. I'm also not sure why the behavior is like that but I think we can skip that part to not spend too much time on it.
What do you think about such code then?


```go
       // ...
	configOverrides := &clientcmd.ConfigOverrides{CurrentContext: loadKubecontext("")}
	return transform(clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides))
}

func loadKubecontext(fallbackValue string) string {
	//  1. --kubecontext flag
	if kubecontext != "" {
		return kubecontext
	}

	// 2. KUBECONTEXT env
	kubeCtx := os.Getenv("KUBECONTEXT")
	if kubeCtx != "" {
		return kubeCtx
	}

	return fallbackValue
}

func transform(c clientcmd.ClientConfig) (*ConfigWithMeta, error) {
	rawConfig, err := c.RawConfig()
	if err != nil {
		return nil, fmt.Errorf("while getting raw config: %v", err)
	}

	clientConfig, err := c.ClientConfig()
	if err != nil {
		return nil, fmt.Errorf("while getting client config: %v", err)
	}
	
	return &ConfigWithMeta{
		K8s:            clientConfig,
		CurrentContext: loadKubecontext(rawConfig.CurrentContext),
	}, nil
}

```

Does it look a bit cleaner? If so, then please apply the snippet 👍

Reply:
Good by me :+1: .